composer/datasets/cifar10.py

```diff
@@ -1,60 +1,80 @@
 # Copyright 2021 MosaicML. All Rights Reserved.
 
+import textwrap
 from dataclasses import dataclass
+from typing import Optional
 
+import torch.utils.data
 import yahp as hp
 from torchvision import transforms
 from torchvision.datasets import CIFAR10
 
 from composer.core.types import DataLoader
 from composer.datasets.dataloader import DataloaderHparams
-from composer.datasets.hparams import DatasetHparams
+from composer.datasets.hparams import (DatadirHparamsMixin, DatasetHparams, DropLastHparamsMixin, IsTrainHparamsMixin,
+                                       NumTotalBatchesHparamsMixin, ShuffleHparamsMixin, SyntheticBatchesHparamsMixin)
+from composer.datasets.synthetic import SyntheticBatchPairDatasetHparams
 from composer.utils import ddp
 
 
 @dataclass
-class CIFAR10DatasetHparams(DatasetHparams):
+class CIFAR10DatasetHparams(DatasetHparams, ShuffleHparamsMixin, DropLastHparamsMixin, DatadirHparamsMixin,
```

Comment:
The Mixin defines the hparam fields so it helps reduce duplciate code. In addition, it works nicely with pyright. A previous alternative (https://github.com/mosaicml/composer/pull/110/commits/9750feff9bc4101bef91313a8c514c6401d4379b) resulted in somewhat messy getters / setters on the dataset hparams class, plus lots of duplicated fields across datasets.

For readability, I can have just one mixin that combines them all?

Reply:
Discussed offline:
* Keep the synthetic mixin as is
* Move the other mixins into the dataset hparams parent
* Update the synthetic dataset to be infinitely sized
* Move the `torch.utils.data.Subset` wrapper into the `dataloader.initialize_object`, and have this function take `max_len: Optional[int]`. Raise an error somewhere if max_len is None and the dataset type is Synthetic (since otherwise synthetic would never finish the epoch)

Reply:
Addressed the first two points, but couldn't do the last one, since the sampler must be created on the subset dataset.